### PR TITLE
Feat/LIVE-9513 config lib dev

### PR DESCRIPTION
**Pull request by hzheng-ledger**
### 📝 Description

1. Add GetValueByKey helper in live-config lib.
2. Default values for GetValueByKey
3. Unit tests for GetValueByKey.
4. live-config lib refactoring
5. Multiple providers supported in live-config lib.

### ❓ Context

- **JIRA or GitHub link**: https://ledgerhq.atlassian.net/browse/LIVE-9513

### ✅ Checklist

Pull Requests must pass the CI and be code reviewed. Set as Draft if the PR is not ready.

- [X] `npx changeset` was attached.
- [X] **Covered by automatic tests.** 
- [X] **Impact of the changes:** No impact in this PR. Only non reg test are required.
  - 

---

### 🧐 Checklist for the PR Reviewers

- [ ] **The code aligns with the requirements** described in the linked JIRA or GitHub issue.
- [ ] **The PR description clearly documents the changes** made and explains any technical trade-offs or design decisions.
- [ ] **There are no undocumented trade-offs**, technical debt, or maintainability issues.
- [ ] **The PR has been tested** thoroughly, and any potential edge cases have been considered and handled.
- [ ] **Any new dependencies** have been justified and documented.
- [ ] **Performance** considerations have been taken into account. (changes have been profiled or benchmarked if necessary)

### CODE REVIEW

File: apps/ledger-live-desktop/src/config/configInit.ts
Comment (hedi-edelbloute) [inline apps/ledger-live-desktop/src/config/configInit.ts:7]: shouldn't this be development ? (I didn't check node_env value)

File: libs/ledger-live-common/src/config/sharedConfig.ts
Comment (hedi-edelbloute) [inline libs/ledger-live-common/src/config/sharedConfig.ts:3]: We could divide this config in multiple types, for instance this schema could be CosmosConfig and we extend it here ? just an idea

So cosmos logic stays in cosmos folder

File: libs/live-config/src/providers/firebaseRemoteConfig/index.ts
Comment (hedi-edelbloute) [inline libs/live-config/src/providers/firebaseRemoteConfig/index.ts:19]: Can we unit test main methods ? It serves as comment aswell

**Review by hedi-edelbloute — COMMENTED** [review]
Weren't we supposed to have a generic LiveConfig<T> ?

File: apps/ledger-live-desktop/src/config/configInit.ts
Comment (hzheng-ledger) [reply apps/ledger-live-desktop/src/config/configInit.ts:7]: I am not following what you meant ?
process.env.NODE_ENV can be "test", "development" or "production".

File: libs/ledger-live-common/src/config/sharedConfig.ts
Comment (hzheng-ledger) [reply libs/ledger-live-common/src/config/sharedConfig.ts:3]: as this config variable "liveConfig" here is a global variable for all the config in ledger live. (It contains not only the cosmos family config, but other configs in the future)
So I type it as ConfigSchema. 

to inject this, we do this when the app starts:
LiveConfig.setConfig(liveConfig);

if we want to get a strong typed config as CosmosConfig.
we can do that on client side like this:
const cosmosConfig =  LiveConfig.getValueBykey("cosmos") as CosmosConfig;

if we move this config to the cosmos folder, the common part (where we did "LiveConfig.setConfig(liveConfig);") will refer to the cosmos part(inside the families folder) which is an anti pattern.

File: libs/live-config/src/providers/firebaseRemoteConfig/index.ts
Comment (hzheng-ledger) [reply libs/live-config/src/providers/firebaseRemoteConfig/index.ts:19]: we have unit test for the jsonFileReader provider getValueBykey method in "/libs/live-config/src/__tests__/LiveConfig.test.ts". 
The idea is that the LiveConfig can plug any provider as long as it implements the "getValueBykey" method.
It is not possible to write unit test for firebase remote config provider because the it queries the firebase server.

File: apps/ledger-live-desktop/src/config/configInit.ts
Comment (hedi-edelbloute) [reply apps/ledger-live-desktop/src/config/configInit.ts:7]: yes, you wrote "developement" here :p

Comment (hzheng-ledger) [reply apps/ledger-live-desktop/src/config/configInit.ts]: very good catch :P 
fixed now

File: libs/ledger-live-common/src/config/sharedConfig.ts
Comment (hedi-edelbloute) [reply libs/ledger-live-common/src/config/sharedConfig.ts:3]: import ConfigCosmos from 'families/cosmos/config.ts"
ConfigLLC: {
cosmos: ConfigCosmos
}

File: libs/live-config/src/providers/firebaseRemoteConfig/index.ts
Comment (hedi-edelbloute) [reply libs/live-config/src/providers/firebaseRemoteConfig/index.ts:19]: it("should return default value if provider is undefined")
it("should return default value if parser returns null")

Comment (hzheng-ledger) [reply libs/live-config/src/providers/firebaseRemoteConfig/index.ts:21]: fixed

File: libs/ledger-live-common/src/config/sharedConfig.ts
Comment (hzheng-ledger) [reply libs/ledger-live-common/src/config/sharedConfig.ts:3]: fixed in my another PR:
https://github.com/LedgerHQ/ledger-live/pull/5763/files#diff-32a5530819ada406eb6bf26e05e32b0e3cd028eec5d59615726233eeebccb3b8R6

Comment (hedi-edelbloute) [reply libs/ledger-live-common/src/config/sharedConfig.ts:3]: Ok

File: libs/ledger-live-common/src/config/sharedConfig.ts
Comment (hedi-edelbloute) [inline libs/ledger-live-common/src/config/sharedConfig.ts:22]: is it me or there are extra spaces before some keys ? maybe run prettier

File: libs/ledger-live-common/src/config/sharedConfig.ts
Comment (hedi-edelbloute) [inline libs/ledger-live-common/src/config/sharedConfig.ts:70]: We miss a ledgerValidator here /!\

https://ledger-live-production-default-rtdb.europe-west1.firebasedatabase.app/config.json

Please check all of them

File: libs/ledger-live-common/.unimportedrc.json
Comment (hedi-edelbloute) [inline libs/ledger-live-common/.unimportedrc.json:293]: Is it actually intended ? Maybe it's because we import this on LLD and LLM only

File: libs/live-config/src/providers/firebaseRemoteConfig/parser.ts
Comment (hedi-edelbloute) [inline libs/live-config/src/providers/firebaseRemoteConfig/parser.ts:9]: Why do we have to access it with ["type"] ?

File: libs/live-config/src/providers/jsonFileReader/index.ts
Comment (hedi-edelbloute) [inline libs/live-config/src/providers/jsonFileReader/index.ts:30]: Shouldn't we manage the case where info is undefined/null ? Just a question

Accessing it directly typescript will warn us, but if we do something dynamic, for example accessing a family or currency dynamically, i'm not sure it will be fine

Maybe we should throw if info is null/undefined ?

File: libs/live-config/src/providers/jsonFileReader/index.ts
Comment (hedi-edelbloute) [inline libs/live-config/src/providers/jsonFileReader/index.ts:16]: Shouldn't we use an async method ?

Using a sync method here could cause performance issues

Also I wonder if we really want to use fs, isn't it wrong if we port the app in the web ? maybe just a http request for fetching could work

File: libs/live-config/src/providers/jsonFileReader/index.ts
Comment (hedi-edelbloute) [inline libs/live-config/src/providers/jsonFileReader/index.ts:26]: We could also log this if JSON.parse fails

Maybe separate two cases : key doesn't exist, parsing error

Minor

**Review by hedi-edelbloute — COMMENTED** [review]
IMO not shippable as is because of the sync method, rest is fine

File: libs/ledger-live-common/.unimportedrc.json
Comment (hzheng-ledger) [reply libs/ledger-live-common/.unimportedrc.json:293]: yes, as we import this on LLD and LLM.
I found that if I don't add this file here. There would be an CI error.

File: libs/live-config/src/providers/jsonFileReader/index.ts
Comment (hzheng-ledger) [reply libs/live-config/src/providers/jsonFileReader/index.ts:16]: This "jsonFileReader" will not be used in production. This is just an example to show how we add a new tailored provider in this lib.

And the firebase provider will be used in production. (In ledger-live/libs/live-config/src/providers/firebaseRemoteConfig/index.ts) There is no performance issues as the getValue won't fetch config from remote server. It just fetch config from local cache. In our current codebase, the "getValue" method is not async either.

For reference, here is the codebase of firebase SDK. As you can see, the design of "getValue" method is not supposed to be async.
https://github.com/firebase/firebase-js-sdk/blob/537d39982f90aff50519b8c1ad6d58048fb8f244/packages/remote-config/src/api.ts#L209

File: libs/live-config/src/providers/firebaseRemoteConfig/parser.ts
Comment (hzheng-ledger) [reply libs/live-config/src/providers/firebaseRemoteConfig/parser.ts:9]: As we have 
```
type ConfigInfoShape<Type extends keyof ValidConfigTypes> = {
  type: Type;
  default: ValidConfigTypes[Type];
};

export type ConfigInfo =
  | ConfigInfoShape<"string">
  | ConfigInfoShape<"boolean">
  | ConfigInfoShape<"number">
  | ConfigInfoShape<"object">
  | ConfigInfoShape<"array">;

```

the ConfigInfo["type"] accepts "string", "boolean", "number", "array" and "object". Do you have any suggestion?

File: libs/ledger-live-common/src/config/sharedConfig.ts
Comment (hzheng-ledger) [reply libs/ledger-live-common/src/config/sharedConfig.ts:70]: good catch. In fact, I miss this config because the "SharedConfig" is a migration of the current codebase from: https://github.com/LedgerHQ/ledger-live/blob/24ebd63bf5c6b9534e4d6f103736f2bfc2d47df5/libs/ledger-live-common/src/config/defaultConfig.ts
I now add the missing injective config in my another PR: 
https://github.com/LedgerHQ/ledger-live/pull/5763/commits/3e4d8cdb7644737b0b5d4079e99e3320d9265561

File: libs/ledger-live-common/src/config/sharedConfig.ts
Comment (hzheng-ledger) [reply libs/ledger-live-common/src/config/sharedConfig.ts:22]: it is weird. I don't find the "extra spaces" that you mentioned. 
by the way, this file is not longer used after my another PR https://github.com/LedgerHQ/ledger-live/pull/5763/commits/

I run the ```pnpm prettier ``` on my side and it seems that everything is OK

File: libs/live-config/src/providers/jsonFileReader/index.ts
Comment (hzheng-ledger) [reply libs/live-config/src/providers/jsonFileReader/index.ts:26]: I think it doesn't matter a lot as the jsonFileReader is not used in production. It is just an example to show how to use this lib. We use only firebaseRemoteConfig as provider in production.

File: libs/live-config/src/providers/jsonFileReader/index.ts
Comment (hzheng-ledger) [reply libs/live-config/src/providers/jsonFileReader/index.ts:30]: the jsonFileReader is not used in production. This is just an example to show how we add a new tailored provider in this lib.

We can handle the case for firebase provider when info is null/undefined 
please refer to
 /libs/live-config/src/providers/firebaseRemoteConfig/index.ts
and 
libs/live-config/src/providers/firebaseRemoteConfig/parser.ts

in the case of firebase provider that we use in production, when info is null/undefined, it means that when we use setConfig from our client. The injected config doesn't include that key. this is the case of featureflag. In my implementation, the setConfig only inject the cosmos family related config. So, when info is null/undefined, I don't throw an exception, instead, I handle it as an object and parse it with JSON.parse (see  /libs/live-config/src/providers/firebaseRemoteConfig/index.ts)

File: libs/live-config/src/featureFlags/firebaseFeatureFlags.ts
Comment (haammar-ledger) [inline libs/live-config/src/featureFlags/firebaseFeatureFlags.ts:12]: I don't think the ! is needed on `configInstance!` you should remove it if possible (also on other places in this file)

File: libs/live-config/src/LiveConfig.ts
Comment (haammar-ledger) [inline libs/live-config/src/LiveConfig.ts:31]:
```suggestion
  public static instance: LiveConfig = new LiveConfig();
```

File: libs/live-config/src/providers/firebaseRemoteConfig/parser.ts
Comment (haammar-ledger) [inline libs/live-config/src/providers/firebaseRemoteConfig/parser.ts:4]:
```suggestion
const objectParser = (value: Value): object | undefined => {
```

File: libs/live-config/src/providers/firebaseRemoteConfig/parser.ts
Comment (haammar-ledger) [inline libs/live-config/src/providers/firebaseRemoteConfig/parser.ts:14]: Since you only use value as Value, you can make the assumption in the param type directly:
```suggestion
export function parser(value: Value, type: ConfigInfo["type"]) {
```

File: libs/live-config/src/providers/index.ts
Comment (hzheng-ledger) [reply libs/live-config/src/providers/index.ts]: good point. Fixed.

File: libs/live-config/src/featureFlags/firebaseFeatureFlags.ts
Comment (hzheng-ledger) [reply libs/live-config/src/featureFlags/firebaseFeatureFlags.ts]: good point. Fixed.

File: libs/live-config/src/LiveConfig.ts
Comment (hzheng-ledger) [reply libs/live-config/src/LiveConfig.ts]: fixed

File: libs/live-config/src/providers/firebaseRemoteConfig/parser.ts
Comment (hzheng-ledger) [reply libs/live-config/src/providers/firebaseRemoteConfig/parser.ts]: good point. Fixed.

File: libs/live-config/src/providers/firebaseRemoteConfig/parser.ts
Comment (hzheng-ledger) [reply libs/live-config/src/providers/firebaseRemoteConfig/parser.ts]: fixed

File: libs/live-config/src/providers/jsonFileReader/index.ts
Comment (hedi-edelbloute) [reply libs/live-config/src/providers/jsonFileReader/index.ts:16]: If this provider isn't used why don't we just add how to add a provider in the documentation ? I think we should not keep unused code, WDYT ?

Comment (hzheng-ledger) [reply libs/live-config/src/providers/jsonFileReader/index.ts:16]: yes, we can remove it of course. but it is used in the unit tests in live-config lib. This is why we keep it here.
we can't use the firebase provider in unit tests directly as it depends on remove server. so we use this simple provider instead.

File: libs/live-config/src/providers/firebaseRemoteConfig/parser.ts
Comment (hedi-edelbloute) [reply libs/live-config/src/providers/firebaseRemoteConfig/parser.ts]: Oops indeed it's fine

File: libs/live-config/src/providers/jsonFileReader/index.ts
Comment (hedi-edelbloute) [reply libs/live-config/src/providers/jsonFileReader/index.ts:30]: Only used as a unit test

File: libs/live-config/src/providers/jsonFileReader/index.ts
Comment (hedi-edelbloute) [reply libs/live-config/src/providers/jsonFileReader/index.ts:16]: Will only be used in unit tests

File: libs/live-config/src/providers/jsonFileReader/index.ts
Comment (hedi-edelbloute) [reply libs/live-config/src/providers/jsonFileReader/index.ts:26]: Will only be used in unit tests

**Review by hedi-edelbloute — DISMISSED** [review]
lgtm when we will move jsonFileReader

**Comment by hzheng-ledger** [comment]
> lgtm when we will move jsonFileReader

jsonFileReader is removed. Thanks

**Review by haammar-ledger — DISMISSED** [review]
Since this is not awaited for a specific release, let's try to merge _just after_ the start of next pre-release, so we have at least a few days to fix any unexpected impact.

**Review by haammar-ledger — APPROVED** [review]

**Comment by gre** [comment]
catching up after this merge. it seems that some code is moving back to live-common? it could have been better to make a new library for the feature flag hooks part, as we have a general effort to split things out of live-common. anyway, we'll re-move this later.
please wait before merging when code affects code owners from other team 🙏

**Pull request merged**
